src/toast/io/observation_hdf_load.py

```diff
@@ -0,0 +1,359 @@
+    det_nelem = dist_dets[obs.comm.group_rank].n_elem
+
+    for field in list(hgrp.keys()):
+        if fields is not None and field not in fields:
+            continue
+        ds = hgrp[field]
+        full_shape = ds.shape
+        dtype = ds.dtype
+        units = u.Unit(str(ds.attrs["units"]))
+
+        sample_shape = None
+        if len(full_shape) > 2:
+            sample_shape = full_shape[2:]
+
+        # All processes create their local detector data
+        obs.detdata.create(
+            field,
+            sample_shape=sample_shape,
+            dtype=dtype,
+            detectors=obs.local_detectors,
+            units=units,
+        )
+
+        # All processes independently load their data
+        obs.detdata[field][:] = ds[
+            det_off : det_off + det_nelem, samp_off : samp_off + samp_nelem
+        ]
+        del ds
+
+
+@function_timer
+def load_hdf5_intervals(obs, hgrp, times, fields):
+    for field in list(hgrp.keys()):
+        if fields is not None and field not in fields:
+            continue
+        # The dataset
+        ds = hgrp[field]
+        # Only the root process reads
+        global_times = None
+        if obs.comm.group_rank == 0:
+            global_times = np.transpose(ds[:])
+
+        obs.intervals.create(field, global_times, times, fromrank=0)
+        del ds
+
+
+# FIXME:  Add options here to prune detectors on load.
+
+
+@function_timer
+def load_hdf5(
+    path,
+    comm,
+    process_rows=None,
+    meta=None,
+    detdata=None,
+    shared=None,
+    intervals=None,
+):
+    """Load an HDF5 observation.
+
+    By default, all detdata, shared, intervals, and noise models are loaded into
+    memory.  A subset of objects may be specified with a list of names
+    passed to the corresponding function arguments.
+
+    Args:
+        path (str):  The path to the file on disk.
+        comm (toast.Comm):  The toast communicator to use.
+        process_rows (int):  (Optional) The size of the rectangular process grid
+            in the detector direction.  This number must evenly divide into the size of
+            comm.  If not specified, defaults to the size of the communicator.
+        meta (list):  Only save this list of metadata objects.
+        detdata (list):  Only save this list of detdata objects.
+        shared (list):  Only save this list of shared objects.
+        intervals (list):  Only save this list of intervals objects.
+
+    Returns:
+        (Observation):  The constructed observation.
+
+    """
+    log = Logger.get()
+    env = Environment.get()
+    parallel = have_hdf5_parallel()
+
+    # Open the file and get the root group.  In both serial and parallel HDF5,
+    # multiple readers are supported.  We open the file with all processes to
+    # enable reading detector and shared data in parallel.
+    hf = None
+    hfgroup = None
+    file_version = None
+    if parallel:
+        hf = h5py.File(path, "r", driver="mpio", comm=comm.comm_group)
```

Comment:
Is there any advantage to opening the file in the parallel access mode when only reading?

Reply:
My "hope" was that using the parallel driver for reading would somehow result in fewer filesystem accesses when getting the HDF5 file information (hierarchy of groups, etc).  But I don't know what HDF5 is doing internally, so I am not sure if it is loading that on one process and broadcasting, etc.